[FSDP2][BE] Refactored `check_1d_sharded_parity` to use mesh

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121660
* __->__ #121357
* #121360

Eventually, we should just have one unified way to check for parity between a `DTensor`-sharded model and a replicated model. This PR is a small refactor to work toward that. One current gap to use this `check_sharded_parity` function for 2D is that FSDP's `(Shard(0), Shard(0))` layout differs from that of the `DTensor` APIs since FSDP shards on dim-0 after TP shards on dim-0.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
[FSDP2] Refactored `check_1d_sharded_parity` to use mesh

[ghstack-poisoned]

Commit:
Update on "[FSDP2][BE] Refactored `check_1d_sharded_parity` to use mesh"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2][BE] Refactored `check_1d_sharded_parity` to use mesh"


Eventually, we should just have one unified way to check for parity between a `DTensor`-sharded model and a replicated model. This PR is a small refactor to work toward that. One current gap to use this `check_sharded_parity` function for 2D is that FSDP's `(Shard(0), Shard(0))` layout differs from that of the `DTensor` APIs since FSDP shards on dim-0 after TP shards on dim-0.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -976,18 +972,22 @@ def check_1d_sharded_parity(
-        if not check_grads:
-            continue
+        mesh, placements = sharded_param.device_mesh, sharded_param.placements
+        if tuple(placements) == (Shard(0), Shard(0)):
```
We will support this in the future with strided sharding. For now, FSDP's `(Shard(0), Shard(0))` differs from `distribute_tensor()`'s `(Shard(0), Shard(0))` since FSDP's one shards dim-1 (TP) first then dim-0, whereas `distribute_tensor()` shards dim-0 then dim-1.

weifengpy reviewed (approved):
nice! the parity checking is flexiable now

awgu commented:
@pytorchbot merge

Commit:
Update on "[FSDP2][BE] Refactored `check_1d_sharded_parity` to use mesh"


Eventually, we should just have one unified way to check for parity between a `DTensor`-sharded model and a replicated model. This PR is a small refactor to work toward that. One current gap to use this `check_sharded_parity` function for 2D is that FSDP's `(Shard(0), Shard(0))` layout differs from that of the `DTensor` APIs since FSDP shards on dim-0 after TP shards on dim-0.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
